Extended support for in-line chained lambdas

SougandhS opened the pull request:
This pr enables users to select a lambda in a single-line chained lambda expressions to put a lambda entry breakpoint + updated chained lambda breakpoint names in breakpoints view 


Fixes : https://github.com/eclipse-jdt/eclipse.jdt.debug/issues/732

On toggling lambda entry breakpoint on chained lambda exps 
<img width="446" alt="image" src="https://github.com/user-attachments/assets/af9a4a73-df20-45a9-9f6f-eac8818a49dd" />

Debugger will show an UI to select the indented lambda of user's choice

<img width="560" height="500" alt="image" src="https://github.com/user-attachments/assets/e5c97101-3783-4d87-ba84-f8444ae7f2e7" />

User can select the lambda inside the small popup either by double click or click->select via button,

On selecting, breakpoint will be added specific to that lambda

<img width="475" height="469" alt="image" src="https://github.com/user-attachments/assets/dd256301-1f78-4a7b-bba3-339fb255e83f" />





https://github.com/user-attachments/assets/3ca9cf35-4528-40a9-a40c-86319b949c5c




<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/eclipse-jdt/.github/blob/main/CONTRIBUTING.md

Note: Security vulnerabilities should not be disclosed on GitHub, through a PR or any
other means. See https://github.com/eclipse-jdt/.github/security/policy
-->

## What it does
<!-- Include relevant issues and describe how they are addressed. -->

## How to test
<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify performance improvements. -->

## Author checklist

- [x] I have thoroughly tested my changes
- [x] The change is following the [coding conventions](https://wiki.eclipse.org/Platform/How_to_Contribute#Coding_Conventions)
- [x] I have signed the [Eclipse Contributor Agreement (ECA)](https://www.eclipse.org/legal/ECA.php)

SougandhS commented:
@iloveeclipse - Could you please review when you get time ?

SougandhS commented:
random Test failure -> https://github.com/eclipse-jdt/eclipse.jdt.debug/issues/719

iloveeclipse reviewed (changes requested):
Few impressions after a quick black box test, without looking into the code.
In general, the idea is useful, but implementation has at least one major bug and also UI needs to be polished.

Take the file below and try that below:

```
import java.util.Arrays;
import java.util.List;
import java.util.Optional;

public class BreakpointsTest {
	
	public static void main(String[] args) {
		callMe();
	}

	private static void callMe() { 
		BreakpointsTest t = new BreakpointsTest();
		t.lotOfLambdas();
	}
	
	void lotOfLambdas() {
		Optional.of("aa").map(this::a6).filter(x -> x.isBlank()).map(this::a6);

		List<String> list = Arrays.asList("a", "b", "c");
		list.stream().filter(s -> s.length() > 0).map(s -> s.toLowerCase()).filter(x -> x.charAt(0) > 1).map(s -> s.toUpperCase()).filter(k -> k.indexOf(0) < 0).map(s -> s.toLowerCase()).filter(p -> !p.isBlank()).forEach(System.out::println);
		
		list.stream().filter(s -> s.length() > 0).map(s -> s.toLowerCase()).filter(x -> x.charAt(0) > 1).map(s -> s.toUpperCase())
				.filter(k -> k.indexOf(0) < 0).map(s -> s.toLowerCase()).filter(p -> !p.isBlank()).forEach(System.out::println);
		
		list.stream().filter(s -> s.length() > 0).map(s -> s.toLowerCase()).filter(x -> x.charAt(0) > 1)
				.map(s -> s.toUpperCase()).filter(k -> {
					System.out.println(k);
					return k.indexOf(0) < 0;
				}).map(s -> s.toLowerCase()).filter(p -> !p.isBlank()).forEach(t -> {
					if(t.contains("a")) {
						System.out.println(t + "!!!");
					}
				});
	}
	
	String a6(String s) {
		return s;
	}
}

```

1. Major bug: "Toggle Breakpoint" works to create breakpoints but is broken for deleting breakpoints, breakpoints can't be removed from editor anymore (only via Breakpoints view). It seem breakpoints can be "toggled" only on lines where lambda breakpoints can be created via "Toggle Lambda Breakpoint" menu.
2. Bug: if one closes lambda proposal dialog by "cancel", no breakpoint is created **but** on next double click on same line no lambda proposals dialog is shown and a regular breakpoint is created. Expectation is: on "Cancel" nothing is created and no behavior is changed. After playing with that for some time I was not able to show up lambda selection dialog anymore, independently what I've tried - **without any error in the log**.
3. UI problem: the shown list of lambda's is resizable but the content inside is fixed.
4. UI problem: extra line added for block expressions: 
![image](https://github.com/user-attachments/assets/0f6c47fd-bef0-42ed-a95e-d15b34a040d1)

SougandhS commented:
> Few impressions after a quick black box test, without looking into the code. In general, the idea is useful, but implementation has at least one major bug and also UI needs to be polished.

Thanks for testing, will fix and get back 👍

SougandhS commented:
> 1. Major bug: 

Fixed
> 2.  no breakpoint is created but on next double click on same line no lambda proposals dialog is shown

Lambda proposal is only shown when user do **Right click->Toggle lambda Entry breakpoint** . Double click creates Line breakpoint only

>After playing with that for some time I was not able to show up lambda selection dialog

Strange. I tried many times and it was working for me, can you try again ? or share the exact same step you tried

> 3 the shown list of lambda's is resizable but the content inside is fixed.

Fixed

> 4 extra line added for block expressions:

Handled

SougandhS commented:
Hi @iloveeclipse 
Could you please re-review this once you are available ?

SougandhS commented:
Hi @iloveeclipse 
I’ve made some code optimizations and fixed a few UI issues. Could we plan this for M2 or M3?

iloveeclipse commented:
I hope to find time tomorrow

iloveeclipse commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/actions/ActionMessages.properties:
```diff
@@ -154,4 +154,7 @@ Override_Dependencies_title=Override Dependencies
-Override_Dependencies_label2=Dependencies for launching:
\ No newline at end of file
+Override_Dependencies_label2=Dependencies for launching:
+lambdaSelection=Select a lambda
```
I believe `Select lambda` is better.

iloveeclipse commented on org.eclipse.jdt.debug/model/org/eclipse/jdt/debug/core/IJavaMethodBreakpoint.java:
```diff
@@ -156,4 +156,58 @@ public interface IJavaMethodBreakpoint extends IJavaLineBreakpoint {
+	/**
+	 * Returns whether this breakpoint a lambda entry breakpoint or not.
+	 *
+	 * @return <code>true</code> if this breakpoint of lambda entry breakpoint; <code>false</code> if this breakpoint not lambda entry breakpoint;
```
please change the end to ` <code>false</code> if not.`

iloveeclipse commented on org.eclipse.jdt.debug/model/org/eclipse/jdt/internal/debug/core/breakpoints/JavaMethodBreakpoint.java:
```diff
@@ -634,6 +667,13 @@ public void setMarker(IMarker marker) throws CoreException {
 						JDIDebugBreakpointMessages.JavaMethodBreakpoint_0, e));
 			}
 		}
+		boolean checkIfLambda = marker.getAttribute(LAMBDA_BREAKPOINT, false);
```
why do we need extra variable `checkIfLambda`, why not assign directly to `isLambdaBreakpoint`?

iloveeclipse commented on org.eclipse.jdt.debug/model/org/eclipse/jdt/internal/debug/core/breakpoints/JavaMethodBreakpoint.java:
```diff
@@ -778,4 +818,58 @@ protected void addInstanceFilter(EventRequest request,
 		}
 	}
+
+	/**
```
Please do not add "empty" javadocs like here. Just wastes space and is zero help for reader.

Here and overall below.

SougandhS replied:
Updated all

iloveeclipse commented on org.eclipse.jdt.debug/model/org/eclipse/jdt/internal/debug/core/breakpoints/LambdaLocationLocator.java:
```diff
@@ -67,10 +73,20 @@ public boolean visit(LambdaExpression node) {
-			fLambdaMethodSignature = LambdaLocationLocatorHelper.toMethodSignature(methodBinding);
-			fLocationFound = true;
+			fCount++;
+			if (fCount == fLambdaPosition || fLambdaPosition == -10) {
```
`-10` is a no go. This seem to be a "special" value - so please add a constant with good name what kind of "special" value it is.

iloveeclipse commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/actions/ToggleBreakpointAdapter.java:
```diff
@@ -448,6 +462,16 @@ private static void doToggleMethodBreakpoint(IMethod member, String lambdaMethod
+			methodBreakpoint.setLambdaBreakpoint(true);
+			methodBreakpoint.setInlineLambdas(lambdaPosition);
+			String lambdaName = lambdaProperties.getLambdaName();
+			if (lambdaProperties.isInline) {
```
`isInline()` please

iloveeclipse commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/actions/ToggleBreakpointAdapter.java:
```diff
@@ -1680,6 +1730,31 @@ private void toggleFieldOrMethodBreakpoints(IWorkbenchPart part, ISelection sele
 		}
 	}
 
+	class LambdaBreakpoint {
```
Can this please be a record?

iloveeclipse commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/actions/ToggleBreakpointAdapter.java:
```diff
@@ -1894,4 +1969,99 @@ private static ICompilationUnit getCompilationUnit(ITextEditor editor) {
+				int ind = table.getSelectionIndex();
+				if (ind >= 0) {
+					selection[0] = ind;
+					popup.dispose();
```
Why not `popup.close();` ? Same below.

SougandhS replied:
updated to `popup.close()`

iloveeclipse commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/actions/ToggleBreakpointAdapter.java:
```diff
@@ -1894,4 +1969,99 @@ private static ICompilationUnit getCompilationUnit(ITextEditor editor) {
+		popup.setLayout(null);
+		Table table = new Table(popup, SWT.BORDER | SWT.V_SCROLL | SWT.H_SCROLL);
+		table.setBounds(10, 10, popup.getBounds().width - 20, popup.getBounds().width - 60);
+		List<String> itemsName = lambdaExps.stream().map(LambdaExpression::toString).toList();
```
Please extract to method `toLambdaEntries()`

iloveeclipse commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/actions/ToggleBreakpointAdapter.java:
```diff
@@ -1894,4 +1969,99 @@ private static ICompilationUnit getCompilationUnit(ITextEditor editor) {
+			public void controlResized(ControlEvent e) {
+				Rectangle clientArea = popup.getClientArea();
+				table.setBounds(10, 10, clientArea.width - 20, clientArea.height - 60);
+				selectButton.setBounds(clientArea.width - 180, clientArea.height - 40, 80, 30);
```
I do not like these hard coded absolute position numbers ` - 90`, ` - 180` etc, also `closeButton.setBounds(170, 180, 80, 30);` above.
This all looks like missing / wrong layout used, ideally no absolute numbers should be used.

SougandhS replied:
I have updated the entire layout now 👍

iloveeclipse reviewed (changes requested):
Functionality wise this looks good now.
Please check individual comments.

SougandhS commented on org.eclipse.jdt.debug/model/org/eclipse/jdt/debug/core/IJavaMethodBreakpoint.java:
```diff
@@ -156,4 +156,58 @@ public interface IJavaMethodBreakpoint extends IJavaLineBreakpoint {
+	/**
+	 * Sets the in-line lambda position
+	 *
+	 * @param pos
```
Yep, Starts with 0, and should be >= 0
Updated the JavaDoc

iloveeclipse commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/actions/ToggleBreakpointAdapter.java:
```diff
@@ -1662,14 +1698,30 @@ private void toggleFieldOrMethodBreakpoints(IWorkbenchPart part, ISelection sele
-					toggleLambdaEntryMethodBreakpoints(part, textSelection, firstLambda.getLambdaMethodName(), firstLambda.getfLambdaMethodSignature());
+					ITextSelection textSelection = new TextSelection(document, selectedLambda.getNodeOffset(), selectedLambda.getNodeLength());
+					LambdaBreakpoint lambdaBp = new LambdaBreakpoint(lambdaExpresions.get(selected).toString(), selected, lambdaExpresions.size() > 1
+							? true
```
Still not readable: ternary expression as method argument

iloveeclipse commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/actions/ToggleBreakpointAdapter.java:
```diff
@@ -1894,4 +1952,110 @@ private static ICompilationUnit getCompilationUnit(ITextEditor editor) {
+		closeButton.setLayoutData(new GridData(SWT.FILL, SWT.CENTER, true, false));
+		closeButton.addListener(SWT.Selection, e -> popup.close());
+		int tableHeight = Math.min(lambdaExps.size(), 10) * table.getItemHeight() + 80;
+		popup.setSize(300, tableHeight);
```
still lot of hard coded values for widget sizes etc. Whole manual creation should be replaced with existing `ElementListSelectionDialog` dialog.

iloveeclipse commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/actions/ToggleBreakpointAdapter.java:
```diff
@@ -1894,4 +1952,110 @@ private static ICompilationUnit getCompilationUnit(ITextEditor editor) {
+	private List<String> toLambdaEntries(List<LambdaExpression> lambdaExps) {
+		List<String> itemsName = lambdaExps.stream().map(LambdaExpression::toString).toList();
+		itemsName = itemsName.stream().map(item -> item.trim().replaceAll("\\n+$", "")).toList(); //$NON-NLS-1$ //$NON-NLS-2$
+		if (Platform.getOS().equals(Platform.OS_MACOSX)) { // If an expression is of multiline then we should shorten it to single line.
```
I still can't understand why anything Mac specific should be used here

iloveeclipse commented on org.eclipse.jdt.debug/model/org/eclipse/jdt/internal/debug/core/breakpoints/LambdaCollector.java:
```diff
@@ -0,0 +1,49 @@
+/*******************************************************************************
```
The class is not needed at all

iloveeclipse reviewed (changes requested):
I've rewrote some of the code and will push an update with my changes in an extra commit.

Please test, if it is OK, we could squash, update commit message and merge. But please first test if I didn't broke anything (I believe not).

PR follows soon.

iloveeclipse commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/BreakpointMarkerUpdater.java:
```diff
@@ -113,15 +113,14 @@ public boolean updateMarker(IMarker marker, IDocument document, Position positio
 		try {
 			ValidBreakpointLocationLocator loc;
-			if (breakpoint instanceof IJavaMethodBreakpoint && ((IJavaMethodBreakpoint) breakpoint).isEntry()) {
+			if (breakpoint instanceof IJavaMethodBreakpoint methodBreakpoint && methodBreakpoint.isEntry()) {
```
Now one can see changes on this file are unneeded

SougandhS commented:
@iloveeclipse I have 2 major concerns here : 

1. The new popup is broken at multiple places 
2. I believe this is less user friendly 

see my further explanations below.

1. The new popup is broken as shown
consider this expression

`list.stream().filter(s -> s.length() > 0).map(s -> s.toLowerCase()).filter(x -> x.charAt(0) > 1).map(s -> s.toUpperCase()).filter(k -> k.indexOf(0) < 0).map(s -> s.toLowerCase()).filter(p -> !p.isBlank()).filter(s -> s.length() > 0).map(s -> s.toLowerCase()).filter(x -> x.charAt(0) > 1).map(s -> s.toUpperCase()).filter(k -> k.indexOf(0) < 0).map(s -> s.toLowerCase()).filter(p -> !p.isBlank()).forEach(System.out::println);`

<br><br>
- I got this exception on selecting a lambda from this new popup 
- Can't even create a normal lambda bp now :(

<img width="630" height="171" alt="image" src="https://github.com/user-attachments/assets/9e01a24c-e6c5-45c1-9c87-47c72814b835" />


```
java.lang.NoSuchMethodError: 'java.lang.String org.eclipse.jdt.internal.debug.ui.actions.ToggleBreakpointAdapter.shortenExpression(java.lang.String)'
	at org.eclipse.jdt.internal.debug.ui.actions.ToggleBreakpointAdapter$LambdaBreakpoint.getLambdaName(ToggleBreakpointAdapter.java:1737)
	at org.eclipse.jdt.internal.debug.ui.actions.ToggleBreakpointAdapter.doToggleMethodBreakpoint(ToggleBreakpointAdapter.java:471)
	at org.eclipse.jdt.internal.debug.ui.actions.ToggleBreakpointAdapter.doToggleLambdaEntryMethodBreakpoints(ToggleBreakpointAdapter.java:331)
	at org.eclipse.jdt.internal.debug.ui.actions.ToggleBreakpointAdapter$4.run(ToggleBreakpointAdapter.java:292)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)

```

2. comments on user friendliness 

    - First, the popup shows this.. now user may be confused to choose which `s -> s.toLowerCase()` ?

    <img width="633" height="492" alt="image" src="https://github.com/user-attachments/assets/8fa2d962-73d5-40f2-8fe9-b4a1977801c4" />


    - Second, popup is not light weight (in contrast my implementation had this designed as a light weight popup)
    <img width="1681" height="939" alt="image" src="https://github.com/user-attachments/assets/befb8d9d-0247-4056-b3fe-ead06dc79baf" />

The idea of the custom popup was for users to quickly select the desired lambda breakpoint from a light weight popup - that will pop up at user's current cursor positon (_which is at near to `Toggle Lambda Entry Breakpoint` so no need to move cursor longer for selection_) and since its a light weight design even if user doesn't want to choose any or by accident click clicked `toggle lambda entry` he/she can either click anywhere outside the popup or click close button to close the popup, and the size of the popup is always dynamic, so it will stay shorter and longer based on list size and Im not sure a filter field is needed at all.

   - Third, sorting order of the popup
     Currently the listing is sorted by alphabetically. In contrast my implementation had sorting by their source positions I know this can be argued in both ways, I think we should figure out a way for both options in preferences

SougandhS commented:
> @iloveeclipse I have 2 major concerns here :
> 
> 1. The new popup is broken at multiple places

Looks like the 1st issue (exception) was due to some workspace issues caused in local. 
@iloveeclipse please ignore the 1st concern

SougandhS commented:
@iloveeclipse 

with this sync with selection feature the custom pop up will be very useful, users find it exciting too :)

https://github.com/user-attachments/assets/0f7c9b0d-6cb4-48c3-bb90-112ff06c54bf

SougandhS commented:
If there are no more changes can this be merged ?

iloveeclipse commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/actions/ToggleBreakpointAdapter.java:
```diff
@@ -1894,4 +1968,133 @@ private static ICompilationUnit getCompilationUnit(ITextEditor editor) {
+				FilteredList filteredList = super.createFilteredList(parent);
+				// Disable default sorting to keep the original order
+				filteredList.setComparator(new LambdaPositionComparator(lambdaNames));
+				filteredList.addSelectionListener(new SelectionAdapter() {
```
Small comment should be given why do we want this listener, I will add that

iloveeclipse commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/actions/ToggleBreakpointAdapter.java:
```diff
@@ -1894,4 +1968,133 @@ private static ICompilationUnit getCompilationUnit(ITextEditor editor) {
+			return count == 0 ? name : name + ZERO_WIDTH_SPACE.repeat(count);
+		}).toList();
+		LambdaLabelProvider lambdaLabelProvider = new LambdaLabelProvider(lambdaExps, lambdaNames);
+		ElementListSelectionDialog dialog = new ElementListSelectionDialog(DebugUIPlugin.getShellForModalDialog(), lambdaLabelProvider) {
```
Fixed

SougandhS commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/actions/ToggleBreakpointAdapter.java:
```diff
@@ -136,7 +149,11 @@
+	private static final String LAMBDA_SEPARATOR = " "; //$NON-NLS-1$
+	private static final int MAX_LAMBDA_LINE_LENGTH = 25;
+	private static final int MAX_TOTAL_LAMBDA_LINE_LENGTH = MAX_LAMBDA_LINE_LENGTH * 2;
+	private static final String ZERO_WIDTH_SPACE = "\u200B"; //$NON-NLS-1$
```
Yes, working 👍

SougandhS commented:
> I will not fix remaining items, squash everyting and update the PR.

PR updated 👍

iloveeclipse commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/BreakpointMarkerUpdater.java:
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2006, 2021 IBM Corporation and others.
+ * Copyright (c) 2006, 2025 IBM Corporation and others.
```
The change should be reverted, it is fully irrelevant to the issue.
